In [2]:
%pip install scanpy 

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 1.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.9/90.9 kB 2.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 35.9 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.9/119.9 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.8/4.8 MB 55.6 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 50.2 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.8/294.8 kB 8.0 MB/s eta 0:00:00ta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 65.1 MB/s eta 0:00:00:00:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.9/233.9 kB 7.0

In [14]:
%pip install MulticoreTSNE

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for MulticoreTSNE: filename=MulticoreTSNE-0.1-cp310-cp310-linux_x86_64.whl size=64702 sha256=48d14556e13d8802380a5c4597a5ffea3d68a6e9b8a5ae4dc2fe845635656cb0
  Stored in directory: /home/codespace/.cache/pip/wheels/59/15/9b/f46b1b5a5e423feb28159b8ff44570b2089e5a53d8eba968e8
Successfully built MulticoreTSNE

[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python3 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd

scseq = pd.read_csv('./data/scRNA_seq_for_hw3_hw5.tsv', delimiter='\t', index_col=0)

In [2]:
scseq.head()

AAACAGCCAAGGAATC-1  AAACAGCCAATCCCTT-1  AAACAGCCAATGCGCT-1  \
HES4                      0.0                 0.0                 0.0   
LINC01342                 0.0                 0.0                 0.0   
TNFRSF18                  0.0                 0.0                 0.0   
TNFRSF4                   0.0                 0.0                 0.0   
AJAP1                     0.0                 0.0                 0.0   

           AAACAGCCACCAACCG-1  AAACAGCCAGGATAAC-1  AAACAGCCAGTAGGTG-1  \
HES4                      0.0                 0.0                 0.0   
LINC01342                 0.0                 0.0                 0.0   
TNFRSF18                  0.0                 0.0                 0.0   
TNFRSF4                   0.0                 0.0                 0.0   
AJAP1                     0.0                 0.0                 0.0   

           AAACAGCCAGTTTACG-1  AAACAGCCATCCAGGT-1  AAACATGCAAGGTCCT-1  \
HES4                      0.0            1.386294                 0.0   
LINC01342                 0.0            0.000000                 0.0   
TNFRSF18                  0.0            0.000000                 0.0   
TNFRSF4                   0.0            0.000000                 0.0   
AJAP1                     0.0            0.000000                 0.0   

           AAACATGCACCGGCTA-1  ...  TTTGTGTTCGGGACCT-1  TTTGTGTTCGTCAAGT-1  \
HES4                 0.693147  ...                 0.0                 0.0   
LINC01342            0.000000  ...                 0.0                 0.0   
TNFRSF18             0.000000  ...                 0.0                 0.0   
TNFRSF4              0.000000  ...                 0.0                 0.0   
AJAP1                0.000000  ...                 0.0                 0.0   

           TTTGTGTTCTCCATAT-1  TTTGTTGGTCAGGAAG-1  TTTGTTGGTCCACAAA-1  \
HES4                      0.0                 0.0                 0.0   
LINC01342                 0.0                 0.0                 0.0   
TNFRSF18                  0.0                 0.0                 0.0   
TNFRSF4                   0.0                 0.0                 0.0   
AJAP1                     0.0                 0.0                 0.0   

           TTTGTTGGTGACATGC-1  TTTGTTGGTGTTAAAC-1  TTTGTTGGTTAGGATT-1  \
HES4                      0.0                 0.0                 0.0   
LINC01342                 0.0                 0.0                 0.0   
TNFRSF18                  0.0                 0.0                 0.0   
TNFRSF4                   0.0                 0.0                 0.0   
AJAP1                     0.0                 0.0                 0.0   

           TTTGTTGGTTGGTTAG-1  TTTGTTGGTTTGCAGA-1  
HES4                      0.0                 0.0  
LINC01342                 0.0                 0.0  
TNFRSF18                  0.0                 0.0  
TNFRSF4                   0.0                 0.0  
AJAP1                     0.0                 0.0  

[5 rows x 10412 columns]

In [3]:
scseq.shape

(3000, 10412)

- each column is obtained via RNA sequencing of a single cell, and can be perceived as an id for each cell.
- each row is the names of the genes that are expressed in the cell. the corresponding value is the level to which the gene is expressed in the cell.
- there are 5 rows and 10412 columns, and therefore this diagram studies the level of expression of these 5 genes in these 10412 cells.

In [4]:
labels = pd.read_csv('./data/label_for_hw3_hw5.tsv', delimiter='\t', index_col=0)

In [5]:
print(labels.head())
print(labels.shape)

                        label
cells                        
AAACAGCCAAGGAATC-1  CD4 Naive
AAACAGCCAATCCCTT-1    CD4 TEM
AAACAGCCAATGCGCT-1  CD4 Naive
AAACAGCCACCAACCG-1  CD8 Naive
AAACAGCCAGGATAAC-1  CD4 Naive
(10412, 1)


here the label is the category or state that the cell is in. this is derived from clustering methods.

- for example, the CD4 Naive, CD4 TEM and CD8 Naive are all cell categories.

In [6]:
import numpy as np
import scanpy as sc
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE
from MulticoreTSNE import MulticoreTSNE as mTSNE
import umap
import seaborn as sns

/usr/local/python/3.10.13/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [7]:
genes =  scseq.T
print(genes.shape)

(10412, 3000)


In [9]:
pca_result = PCA(n_components=2).fit_transform(genes.values)
genes['PCA1'] = pca_result[:,0]
genes['PCA2'] = pca_result[:,1]

In [10]:
genes.shape

(10412, 3002)

In [11]:
# t-SNE
tsne = mTSNE(n_components=2, n_jobs=-1)  # 使用所有可用的CPU核心
tsne_result = tsne.fit_transform(genes.values)
genes['TSNE1'] = tsne_result[:,0]
genes['TSNE2'] = tsne_result[:,1]

In [12]:
print(genes.shape)

(10412, 3004)


In [13]:
# UMAP
umap_result = umap.UMAP().fit_transform(genes.values)
genes['UMAP1'] = umap_result[:,0]
genes['UMAP2'] = umap_result[:,1]

/home/codespace/.local/lib/python3.10/site-packages/sklearn/manifold/_spectral_embedding.py:273: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.
  warnings.warn(


In [14]:
len(genes)

10412

In [15]:
genes['label'] = labels['label'].values.ravel()

In [27]:
# 可视化
fig, axs = plt.subplots(3, figsize=(10, 30))

# PCA
sns.scatterplot(x='PCA1', y='PCA2', hue='label', data=genes, ax=axs[0], palette='viridis')
axs[0].set_title('PCA')

# t-SNE
sns.scatterplot(x='TSNE1', y='TSNE2', hue='label', data=genes, ax=axs[1], palette='viridis')
axs[1].set_title('t-SNE')

# UMAP
sns.scatterplot(x='UMAP1', y='UMAP2', hue='label', data=genes, ax=axs[2], palette='viridis')
axs[2].set_title('UMAP')

plt.savefig('./hw3/figs/vis_hw3_without_scanpy.png')
plt.show()